smarts/sstudio/types.py

```diff
@@ -814,6 +815,51 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
         return shapely_translate(poly, xoff=x, yoff=y)
 
 
+@dataclass(frozen=True)
+class ConfigurableZone(Zone):
+    """A descripter that defines a specific configurableZone defined by user"""
+
+    ext_coordinates: List[Tuple[float, float]]
+    """external coordinates of the polygon
+    < 2 points provided: error
+    = 2 points provided: generates a box using these two points as diagonal
+    > 2 points provided: generates a polygons according to the coordinates"""
+    rotation: Optional[float] = None
+    """The heading direction of the bubble(radians, clock-wise rotation)"""
+
+    def __post_init__(self):
+        if not self.ext_coordinates or not isinstance(self.ext_coordinates[0], tuple):
+            raise ValueError(
+                "Two points or more are needed to create a polygon. (less than two points are provided)"
+            )
+
+        x_set = set(point[0] for point in self.ext_coordinates)
+        y_set = set(point[1] for point in self.ext_coordinates)
+        if len(x_set) == 1 or len(y_set) == 1:
+            raise ValueError(
+                "Parallel line cannot form a polygon. (points provided form a parallel line)"
+            )
```

Comment:
This message is a bit confusing. I think this could be incorporated into the first check above, just add a check for `len(self.ext_coordinates) < 2`.

Reply:
This is to check whether user provides multiple coordinates with identical x_axis or y_axis. This will form a parallel line which is not a polygon.

Reply:
Ah yes, I missed the `set`, sorry. I still think we need an explicit check for len < 2 above though.

Reply:
There is a problem of using `len < 2` to check numbers of coordinates. If the user only provides one coordinate, the ext_coordinate will become something like this: ((90,90)). The `len(ext_coordinate)` should be 2 but python somehow unpacks it to be `ext_coordinate = (90,90)` and its `len = 2` again, this may not be a valid check.

Reply:
Are you sure? As long as it's always a list it should work. From the REPL:

```
>>> len([(90,90)])
1
```

Reply:
oh I tried it with this:
```
test = ((90,90))
print(len(test))
```
and it outputs 2

Reply:
That would be expected, since you are assigning a tuple of length 2. The outside parentheses are ignored.

```
>>> test = ((90,90))
>>> test
(90, 90)
```

I think we can assume that `ext_coordinates` is always a list.

Reply:
Normally the user should input a list, but () also works for multiple coordinates, and `len(((90,90),(110,110))) = 2`. Maybe I can convert it to list if index 0 is not a tuple and do the len check?

Reply:
This should actually work for tuples as well with no change. But the code would be:

```
>>> test = ((90,90),)
>>> len(test)
1
```

Reply:
I see, I will add the len check as well, thanks for your patience